meddlr/metrics/SSFD.py

```diff
@@ -0,0 +1,205 @@
+
+
+class SSFD(Metric):
+    """
+    Self-Supervised Feature Distance. SSFD evaluates the feature distance between a
+    pair of images from features extracted from a pre-trained neural network trained on
+    a self-supervision task with fastMRI datasets [1]. SSFD has been shown to correspond
+    well to Radiologist Reader Scores of accelerated  MR reconstructions.
+
+
+    Attributes:
+        mode (str):
+            This flag determines how to interpret the channel dimension of the inputs.
+            ['grayscale']: Each channel corresponds to a distinct grayscale input image.
+            ['rgb']: The 3 channel dimensions correspond to a single rgb image.
+                     Exception will be thrown if channel dimension != 3.
+
+
+    References:
+    ..  [1] Adamson, Philip M., et al.
+        SSFD: Self-Supervised Feature Distance as an MR Image Reconstruction Quality Metric."
+        NeurIPS 2021 Workshop on Deep Learning and Inverse Problems. 2021.
+        https://openreview.net/forum?id=dgMvTzf6M_3
+    """
+
+    def __init__(
+        self,
+        mode: str = "grayscale",
+        channel_names: Sequence[str] = None,
+        reduction="none",
+        compute_on_step: bool = False,
+        dist_sync_on_step: bool = False,
+        process_group: bool = None,
+        dist_sync_fn: bool = None,
+    ):
+
+        super().__init__(
+            channel_names=channel_names,
+            units="",
+            reduction=reduction,
+            compute_on_step=compute_on_step,
+            dist_sync_on_step=dist_sync_on_step,
+            process_group=process_group,
+            dist_sync_fn=dist_sync_fn,
+        )
+
+        valid_modes = ("grayscale", "rgb")
+        if mode not in valid_modes:
+            raise ValueError(
+                f"Argument `mode` must be one \
+                            of {valid_modes}, but got {mode}."
+            )
+
+        self.mode = mode
+
+        path_manager = env.get_path_manager()
+        file_path = path_manager.get_local_path(SSFD_HUGGINGFACE_URL, force=False)
+
+        self.net = SSFD_Encoder(in_channels=1, out_channels=320)
+        self.net.load_state_dict(torch.load(file_path))
+        self.net.eval()
+
+    def func(self, preds, targets) -> torch.Tensor:
+
+        if self.mode == "grayscale":
+            loss_shape = (targets.shape[0], targets.shape[1])
+
+        elif self.mode == "rgb":
+            if targets.shape[1] != 3:
+                raise ValueError(
+                    f"Channel dimension must have size 3 for rgb mode,\
+                    but got tensor of shape {targets.shape}."
+                )
+
+            loss_shape = (targets.shape[0], 1)
+
+        preds = self.preprocess_ssfd(preds)
+        targets = self.preprocess_ssfd(targets)
+
+        features = {}
+
+        def get_features(name):
+            def hook(model, input, output):
+                features[name] = output.detach()
```

Comment:
when you detach, gradients can't flow through. you may need to reconsider this for training with SSFD